File: Number of Islands/Number_Of_Islands.java
class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0){
            return 0;
        }
        int num = 0;
        int depth = grid.length;
        for (int row = 0; row < depth; row++){
            for (int col = 0; col < grid[row].length; col++){
                if (grid[row][col] == '1'){
                   num += countAndSinkIsland(grid,row,col); 
                }
            }
        }
        return num;
    }
    public int countAndSinkIsland(char[][] grid,int row, int col){
        if (row < 0 || row >= grid.length || col < 0 || col >= grid[row].length || 
           grid[row][col] == '0'){
            return 0;
        }
        grid[row][col] = '0';
        countAndSinkIsland(grid,row + 1,col);
        countAndSinkIsland(grid,row - 1,col);
        countAndSinkIsland(grid,row,col + 1);
        countAndSinkIsland(grid,row,col - 1);
        return 1;
    }
}

File: Coding Problems & My Solutions/Number of Islands/Number_Of_Islands.swift
class Solution {
    func numIslands(_ grid: [[Character]]) -> Int {
        var grid = grid
        var numIslands = 0
        for (i, row) in grid.enumerated() {
            for (j, _) in row.enumerated() {
                if grid[i][j] == "1" {
                    numIslands += 1
                    sinkIsland(&grid, i, j)
                }
            }
        }
        return numIslands
    }
    
    func sinkIsland(_ grid: inout [[Character]], _ i: Int, _ j: Int) {
        if i < 0 || i >= grid.count || j < 0 || j >= grid[i].count {
            return 
        }
        if grid[i][j] == "0" {
            return
        }
        grid[i][j] = "0"
        // Up
        sinkIsland(&grid, i, j - 1)
        // Down
        sinkIsland(&grid, i, j + 1)
        // Left 
        sinkIsland(&grid, i - 1, j)
        // Right
        sinkIsland(&grid, i + 1, j)
    }
}